[docs] Add documentation for Observe (#44816)
expo/expo: docs/pages/eas/observe/reference/metrics.mdx

---

> title: Metrics reference

Author: @Ubax added the metrics from the Notion doc in this file.

> title: Metrics reference

Reviewer: We could maybe add a section in this page with the metric for update downloads.

Author: We're still working on the UI to show these, will add it to the docs once we do.

sidebar_title: Metrics
description: A reference of each performance metric tracked by Expo Observe, including concepts and data handling.
---

Reference for the performance metrics Expo Observe collects, the core concepts used to organize events (sessions and users), and how collected data is retained.

## Concepts

> ### Session

Reviewer 2: Is it session or app session?

My concern here is that in the future we may have more sessions. 

Not necessary for this PR, but something to note for future

Author: This refers to the main session. It is the only one we currently surface on the CLI and the website. I think once we add child sessions we can update this.

> A **session** starts when the app process is launched and ends when the app process is terminated. Each session has a unique identifier and contains all events that occurred during that app launch.

Reviewer 2: A small nit:
Maybe let's use metric here instead of event, as events may be more suitable for the logs

### User

A **user** is identified by an anonymous ID that is unique per app installation. This ID:

- Is generated when the app is first installed
- Persists across app updates
- Is reset if the user uninstalls and reinstalls the app
- Is not Personally Identifiable Information (PII)

This allows you to see metrics across multiple sessions for the same user without collecting personal data.

## Metrics

> All duration metrics are reported in seconds.

Reviewer 2: WDYT about converting this to note?

This metric is collected automatically.

**How to improve:**

- Remove unused native modules.
- Avoid static initializers (`+load` methods in Objective-C, static constructors in C++), and native modules or config plugins that add them.
- Keep the app's memory and CPU usage low so the OS does not kill the process while in the background. This does not affect this metric itself, but makes subsequent launches warm instead of cold.
- If you use `expo-updates` with a non-zero `fallbackToCacheTimeout`, the app launch is blocked waiting for an update check. Keep this value at `0` (default), or set `checkOnLaunch` to `NEVER` or `ERROR_RECOVERY_ONLY` to avoid delaying cold launches.

**Our recommendation:** under 1.5s.

### Warm launch time

> **What it measures:** Unlike a cold launch, the app process already exists. The OS just brings it to the foreground and recreates the view hierarchy. Most native resources and services are already in memory, so this type of launch is significantly faster.

Reviewer 3: [🐧 nit] Minor: this opens with "Unlike a cold launch..." before defining what warm launch *is*. Consider leading with the definition ("a warm launch happens when the OS already has the app process in memory and only needs to bring it back to the foreground") and then contrasting against cold launch — matches the framing of the cold-launch section above.

> Warm launch is largely out of the developer's hands: it is the OS restoring state for a process that is already in memory.

Reviewer 2: nit: when I read this `Warm launch is largely out of the developer's hands`, the first thing which comes to me is that I cannot improve it. Is this the intention here?

Author: I think the idea is that there is no public API that you could call to pre-warm the app and then get more warm launches, this is fully decided by the OS. I've improved the wording there, thanks for the callout!

---
title: Metrics reference
sidebar_title: Metrics

> description: Detailed descriptions of each performance metric tracked by EAS Observe, including concepts and data handling.

Reviewer 4: ```suggestion
description: A reference of each performance metric tracked by EAS Observe, including concepts and data handling.
```

---

> ## Concepts

Reviewer 4: ```suggestion
Reference for the performance metrics EAS Observe collects, the core concepts used to organize events (sessions and users), and how collected data is retained.

## Concepts
```

### Session

A **session** starts when the app process is launched and ends when the app process is terminated. Each session has a unique identifier and contains all events that occurred during that app launch.

### User

A **user** is identified by an anonymous ID that is unique per app installation. This ID:

- Is generated when the app is first installed
- Persists across app updates
- Is reset if the user uninstalls and reinstalls the app

> - Is not personally identifiable information (PII)

Reviewer 4: ```suggestion
- Is not Personally Identifiable Information (PII)
```

This allows you to see metrics across multiple sessions for the same user without collecting personal data.

> ## Metrics

Reviewer 3: I think we could update this section with what I wrote in our internal metrics documentation.

Author: Updated.

> ## Metrics

Reviewer 4: We should also define how "Time" is measured here. Seconds, milliseconds, etc.? I think it's a key piece of information that is missing for clarity.

### Cold launch time

Time to launch the app from scratch when the app's native code and resources are not in memory. This is a native-only metric that is measured before the JavaScript bundle starts executing.

> A cold launch occurs after a fresh install, app upgrade, device reboot, or when the OS has killed the app to reclaim memory. The system must load the app from disk and create the process, making this the slowest launch type.

Reviewer 3: ```suggestion
Time from process creation to when the system finished allocating memory, starting a fresh runtime environment, loading the app's code and resources from disk, and initializing its components before rendering the UI. This is the slowest type of launch and typically occurs after a fresh install, app upgrade, device reboot, or when the OS has killed the app to reclaim memory. It is a native-only metric, meaning that your JavaScript code does not affect this metric, but the React Native runtime initialization is included.
```

Author: Superseded by the copy in the notion doc.

This metric is collected automatically.

### Warm launch time

Time to launch the app when native libraries, app's state, and resources are already cached in memory. This is a native-only metric that is measured before the JavaScript bundle starts executing.

During a warm launch, the system only needs to recreate the UI and restore any preserved state.

This metric is collected automatically.

### Bundle load time

> How long it takes to evaluate your JavaScript bundle. This measures the time the JavaScript runtime spends parsing and executing your bundle code.

Reviewer 4: ```suggestion
Time to evaluate your JavaScript bundle. This measures the time the JavaScript runtime spends parsing and executing your bundle code.
```

Author: Superseded by the copy in the notion doc

This metric is collected automatically.

> ### Time to first render

Reviewer 4: ```suggestion
### Time to first render (TTFR)
```

Time from app launch until your app first renders content to the screen.

This metric is collected automatically when you wrap your root layout with `AppMetricsRoot` (see [Get started](/eas/observe/get-started/)). You can also call `AppMetrics.markFirstRender()` manually if needed.

> ### Time to interactive

Reviewer 4: ```suggestion
### Time to interactive (TTI)
```

> Time from app launch until your app is fully ready for user interaction. This is measured when you call `AppMetrics.markInteractive()`.

Reviewer 3: ```suggestion
Time between the warm/cold launch and when the user can actually tap, scroll, and interact with the app in other ways. It is the most important startup metric because it is what users perceive as "the app is ready".
```

Author: Superseded by the copy in the notion doc.

> **When to call:** After all initialization work is complete and the splash screen is hidden. This should represent the moment when users can actually start using your app.

Reviewer 4: ```suggestion
**When to call:** After all initialization work is complete and the splash screen is hidden. This should represent the moment when users can start using your app.
```

`markInteractive()` can safely be called multiple times per session, only the first call records the measurement. Call it on every screen that users can deep link to. If you only place it on one screen, TTI will be measured incorrectly when the app is opened via a deep link to a different screen: for example, if screen A has the call but the user deep links to screen B, TTI will not be recorded until they navigate to A.

## Data handling

### Offline collection

> Metrics collected while the device is offline are stored locally on the device. They are automatically sent to the server when the app moves to the background and there is connectivity. You can also call `ExpoObserve.dispatchEvents()` to flush events manually at any time (see [Configuration](/eas/observe/configuration/)).

Reviewer 4: ```suggestion
Metrics collected while the device is offline are stored on the device. They are automatically sent to the server when the app moves to the background, provided there is connectivity. You can also call [`ExpoObserve.dispatchEvents()`](/eas/observe/configuration/#dispatchevents) to flush events manually at any time.
```

### Data retention

Metric data is retained for a minimum of 60 days.

### Sampling

All sessions are tracked. There is no sampling applied to the data collection.

### Environment

> All metrics are grouped by environment. The environment value is derived from `process.env.NODE_ENV` by default, or can be overridden via `configure({ environment })`. Metrics with the `development` environment are filtered out before dispatching unless `enableInDebug` is set to `true` in your app config (see [Enable metrics in development](/eas/observe/configuration/#enable-metrics-in-development)). You can also disable all dispatching globally using `configure({ dispatchingEnabled: false })`.

Reviewer 3: We should revisit this before publishing. Let's discuss it on the next meeting.

Author: I suggest we ship this as is and update the docs once @Ubax 's fix lands

> All metrics are grouped by environment. The environment value is derived from `process.env.NODE_ENV` by default, or can be overridden via `configure({ environment })`. Metrics with the `development` environment are filtered out before dispatching unless `enableInDebug` is set to `true` in your app config (see [Enable metrics in development](/eas/observe/configuration/#enable-metrics-in-development)). You can also disable all dispatching globally using `configure({ dispatchingEnabled: false })`.

Reviewer 4: ```suggestion
All metrics are grouped by environment. The environment value is derived from `process.env.NODE_ENV` by default, or can be overridden via [`configure({ environment })`](/eas/observe/configuration/#configure). Metrics with the `development` environment are filtered out before dispatching unless `enableInDebug` is set to `true` in your app config. (for information, see [Enable metrics in development](/eas/observe/configuration/#enable-metrics-in-development)). You can also disable all dispatching globally using `configure({ dispatchingEnabled: false })`.
```
